server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamLifecycle.java

```diff
@@ -198,8 +201,12 @@ public void addWarningHeaderIfDataRetentionNotEffective(@Nullable DataStreamGlob
             return;
         }
         Tuple<TimeValue, DataStreamLifecycle.RetentionSource> effectiveDataRetentionWithSource = getEffectiveDataRetentionWithSource(
-            globalRetention
+            globalRetention,
+            false
         );
+        if (effectiveDataRetentionWithSource == null || effectiveDataRetentionWithSource.v1() == null) {
+            return;
+        }
```

Comment:
1. It seems like this was a bug before this PR, right? It was already possible for `v1()` to return null, which would cause an NPE a couple of lines below. Good catch :)
2. The `effectiveDataRetentionWithSource == null` seems redundant as we always return a tuple in `getEffectiveDataRetentionWithSource()`.

Reply:
I put that check in because `getEffectiveDateRetentionWithSource` is marked as Nullable. But you're right, it's not actually. I might pull this into a separate PR where I also remove the @Nullable annotation.

Reply:
I moved this into https://github.com/elastic/elasticsearch/pull/108159

Reply:
Thanks! :)